lib/puppet/parser/environment_compiler.rb

```diff
@@ -0,0 +1,85 @@
+require 'puppet/parser/compiler'
+
+class Puppet::Parser::EnvironmentCompiler < Puppet::Parser::Compiler
+  def self.compile(env)
+    $env_module_directories = nil
+    env.check_for_reparse
+
+    node = Puppet::Node.new(env)
+    node.environment = env
+    new(node).compile
+  rescue => detail
```

Comment:
I have seen really odd behavior when using a method with rescue - IIRC in conjunction with tests and mocks and chains of rescues. Prefer if the longer being/rescue/end form is used.